vortex-file/src/read/layouts/columnar.rs

```diff
@@ -1,126 +1,85 @@
+                message_cache,
+            }
+            .build()?,
+        ))
     }
 }
 
-/// In memory representation of Columnar NestedLayout.
-///
-/// Each child represents a column
-#[derive(Debug)]
-pub struct ColumnarLayout {
+struct ColumnarLayoutBuilder {
     fb_bytes: Bytes,
     fb_loc: usize,
     scan: Scan,
     layout_serde: LayoutDeserializer,
     message_cache: RelativeLayoutCache,
-    reader: Option<ColumnBatchReader>,
 }
 
-impl ColumnarLayout {
-    pub fn new(
-        fb_bytes: Bytes,
-        fb_loc: usize,
-        scan: Scan,
-        layout_serde: LayoutDeserializer,
-        message_cache: RelativeLayoutCache,
-    ) -> Self {
-        Self {
-            fb_bytes,
-            fb_loc,
-            scan,
-            layout_serde,
-            message_cache,
-            reader: None,
-        }
-    }
-
+impl ColumnarLayoutBuilder {
     fn flatbuffer(&self) -> footer::Layout {
         unsafe {
             let tab = flatbuffers::Table::new(&self.fb_bytes, self.fb_loc);
             footer::Layout::init_from_table(tab)
         }
     }
 
-    /// Perform minimal amount of work to construct children that can be queried for splits
-    fn children_for_splits(&self) -> VortexResult<Vec<Box<dyn LayoutReader>>> {
+    fn build(&self) -> VortexResult<ColumnarLayoutReader> {
         let (refs, lazy_dtype) = self.fields_with_dtypes()?;
         let fb_children = self.flatbuffer().children().unwrap_or_default();
 
-        refs.into_iter()
-            .map(|field| {
-                let resolved_child = lazy_dtype.resolve_field(&field)?;
-                let child_loc = fb_children.get(resolved_child)._tab.loc();
-
-                self.layout_serde.read_layout(
-                    self.fb_bytes.clone(),
-                    child_loc,
-                    Scan::new(None),
-                    self.message_cache.unknown_dtype(resolved_child as u16),
-                )
-            })
-            .collect::<VortexResult<Vec<_>>>()
-    }
-
-    fn column_reader(&self) -> VortexResult<ColumnBatchReader> {
-        let (refs, lazy_dtype) = self.fields_with_dtypes()?;
-        let fb_children = self.flatbuffer().children().unwrap_or_default();
-
-        let filter_dtype = lazy_dtype.value()?;
-        let DType::Struct(s, ..) = filter_dtype else {
-            vortex_bail!("Column layout must have struct dtype")
-        };
-
         let mut unhandled_names = Vec::new();
         let mut unhandled_children = Vec::new();
         let mut handled_children = Vec::new();
         let mut handled_names = Vec::new();
 
-        for (field, (name, dtype)) in refs
-            .into_iter()
-            .zip_eq(s.names().iter().cloned().zip_eq(s.dtypes().iter().cloned()))
-        {
+        for (field, name) in refs.into_iter().zip_eq(lazy_dtype.names()?.iter()) {
             let resolved_child = lazy_dtype.resolve_field(&field)?;
+            let projection = Projection::SingleField(field.clone());
```

Comment:
It seems like this is the reason to add SingleField; right? When we call `names` on line 80, do we deserialize the struct dtype? If yes, it seems like the old code was fine. Are we able to get names without fully deserializing a struct dtype?

Reply:
Though you really want delay fetching names up to when you return the arrays but that can happen later. Ideally we could create layouts eagerly and they would lazily fetch data